File: water_jug_problem.py
"""
Water Jug Problem Solver using DFS and BFS

This module solves the classic water jug problem where we have two jugs
of different capacities and need to measure a specific amount of water.

Problem: Given a 3-liter jug and a 4-liter jug, measure exactly 2 liters
in the 4-liter jug using only fill, empty, and pour operations.
"""

from collections import deque
from typing import List, Tuple, Set, Optional


class WaterJugProblem:
    """
    Solver for the water jug problem using search algorithms.
    
    Attributes:
        jug1_capacity: Maximum capacity of the first jug
        jug2_capacity: Maximum capacity of the second jug
        goal_state: Target state (jug1_amount, jug2_amount)
        initial_state: Starting state (default: (0, 0))
    """
    
    def __init__(self, jug1_capacity: int, jug2_capacity: int, 
                 goal_state: Tuple[int, int], initial_state: Tuple[int, int] = (0, 0)):
        """
        Initialize the water jug problem.
        
        Args:
            jug1_capacity: Capacity of jug 1
            jug2_capacity: Capacity of jug 2
            goal_state: Target state (jug1, jug2)
            initial_state: Initial state (default: (0, 0))
        """
        self.jug1_capacity = jug1_capacity
        self.jug2_capacity = jug2_capacity
        self.goal_state = goal_state
        self.initial_state = initial_state
        self.path_count = 0
        self.max_depth = 20  # Maximum depth for path counting to avoid infinite loops
    
    def get_successors(self, state: Tuple[int, int]) -> List[Tuple[Tuple[int, int], str]]:
        """
        Generate all possible successor states from current state.
        
        Operations:
        1. Fill jug 1
        2. Fill jug 2
        3. Empty jug 1
        4. Empty jug 2
        5. Pour from jug 1 to jug 2
        6. Pour from jug 2 to jug 1
        
        Args:
            state: Current state (jug1, jug2)
            
        Returns:
            List of (next_state, action) tuples
        """
        jug1, jug2 = state
        successors = []
        
        # Operation 1: Fill jug 1
        if jug1 < self.jug1_capacity:
            successors.append(((self.jug1_capacity, jug2), "Fill Jug1"))
        
        # Operation 2: Fill jug 2
        if jug2 < self.jug2_capacity:
            successors.append(((jug1, self.jug2_capacity), "Fill Jug2"))
        
        # Operation 3: Empty jug 1
        if jug1 > 0:
            successors.append(((0, jug2), "Empty Jug1"))
        
        # Operation 4: Empty jug 2
        if jug2 > 0:
            successors.append(((jug1, 0), "Empty Jug2"))
        
        # Operation 5: Pour from jug 1 to jug 2
        if jug1 > 0 and jug2 < self.jug2_capacity:
            pour_amount = min(jug1, self.jug2_capacity - jug2)
            successors.append(((jug1 - pour_amount, jug2 + pour_amount), 
                             "Pour Jug1->Jug2"))
        
        # Operation 6: Pour from jug 2 to jug 1
        if jug2 > 0 and jug1 < self.jug1_capacity:
            pour_amount = min(jug2, self.jug1_capacity - jug1)
            successors.append(((jug1 + pour_amount, jug2 - pour_amount), 
                             "Pour Jug2->Jug1"))
        
        return successors
    
    def dfs(self) -> Optional[List[Tuple[Tuple[int, int], str]]]:
        """
        Solve the water jug problem using Depth First Search.
        
        Returns:
            List of (state, action) tuples representing the solution path,
            or None if no solution exists.
        """
        stack = [(self.initial_state, [])]
        visited = set()
        all_paths = []
        
        while stack:
            state, path = stack.pop()
            
            if state in visited:
                continue
            
            visited.add(state)
            
            # Check if goal is reached
            if state == self.goal_state:
                all_paths.append(path + [(state, "Goal Reached")])
                self.path_count += 1
                if len(all_paths) == 1:  # Return first path found
                    return path + [(state, "Goal Reached")]
            
            # Explore successors
            for next_state, action in self.get_successors(state):
                if next_state not in visited:
                    stack.append((next_state, path + [(state, action)]))
        
        return None
    
    def bfs(self) -> Optional[List[Tuple[Tuple[int, int], str]]]:
        """
        Solve the water jug problem using Breadth First Search.
        
        Returns:
            List of (state, action) tuples representing the solution path,
            or None if no solution exists.
        """
        queue = deque([(self.initial_state, [])])
        visited = set()
        visited.add(self.initial_state)
        all_paths = []
        
        while queue:
            state, path = queue.popleft()
            
            # Check if goal is reached
            if state == self.goal_state:
                all_paths.append(path + [(state, "Goal Reached")])
                self.path_count += 1
                return path + [(state, "Goal Reached")]
            
            # Explore successors
            for next_state, action in self.get_successors(state):
                if next_state not in visited:
                    visited.add(next_state)
                    queue.append((next_state, path + [(state, action)]))
        
        return None
    
    def count_all_paths_dfs(self) -> int:
        """
        Count all possible paths to the goal state using DFS.
        
        Returns:
            Number of different paths to reach the goal state.
        """
        stack = [(self.initial_state, [self.initial_state])]
        visited_paths = set()
        path_count = 0
        
        while stack:
            state, path = stack.pop()
            
            # Check if goal is reached
            if state == self.goal_state:
                path_tuple = tuple(path)
                if path_tuple not in visited_paths:
                    visited_paths.add(path_tuple)
                    path_count += 1
                continue
            
            # Avoid cycles in current path
            if len(path) > self.max_depth:
                continue
            
            # Explore successors
            for next_state, action in self.get_successors(state):
                if next_state not in path:  # Avoid cycles
                    stack.append((next_state, path + [next_state]))
        
        return path_count
    
    def print_solution(self, solution: List[Tuple[Tuple[int, int], str]], 
                      algorithm: str):
        """
        Print the solution path in a readable format.
        
        Args:
            solution: List of (state, action) tuples
            algorithm: Name of the algorithm used
        """
        print(f"\n{'='*60}")
        print(f"Solution using {algorithm}")
        print(f"{'='*60}")
        print(f"Initial State: {self.initial_state}")
        print(f"Goal State: {self.goal_state}")
        print(f"Jug Capacities: Jug1={self.jug1_capacity}L, Jug2={self.jug2_capacity}L")
        print(f"\nSteps to reach goal:")
        print(f"{'Step':<6} {'Action':<20} {'Jug1':<8} {'Jug2':<8}")
        print("-" * 60)
        
        print(f"{'0':<6} {'Initial State':<20} {self.initial_state[0]:<8} {self.initial_state[1]:<8}")
        
        for i, (state, action) in enumerate(solution, 1):
            print(f"{i:<6} {action:<20} {state[0]:<8} {state[1]:<8}")
        
        print(f"\nTotal steps: {len(solution)}")


def main():
    """Main function to demonstrate the water jug problem solver."""
    
    # Problem 1: Original problem (3L, 4L jugs, goal: 2L in 4L jug)
    print("\n" + "="*60)
    print("WATER JUG PROBLEM SOLVER")
    print("="*60)
    
    print("\nProblem 1: 3L and 4L jugs, Goal: (0, 2)")
    print("-" * 60)
    
    problem1 = WaterJugProblem(jug1_capacity=3, jug2_capacity=4, 
                                goal_state=(0, 2), initial_state=(0, 0))
    
    # Solve using DFS
    print("\n[1] Solving with Depth First Search (DFS)...")
    solution_dfs = problem1.dfs()
    if solution_dfs:
        problem1.print_solution(solution_dfs, "DFS")
    else:
        print("No solution found using DFS")
    
    # Solve using BFS
    print("\n[2] Solving with Breadth First Search (BFS)...")
    problem2 = WaterJugProblem(jug1_capacity=3, jug2_capacity=4, 
                                goal_state=(0, 2), initial_state=(0, 0))
    solution_bfs = problem2.bfs()
    if solution_bfs:
        problem2.print_solution(solution_bfs, "BFS")
    else:
        print("No solution found using BFS")
    
    # Count all paths
    print("\n[3] Counting all possible paths to goal state...")
    problem3 = WaterJugProblem(jug1_capacity=3, jug2_capacity=4, 
                                goal_state=(0, 2), initial_state=(0, 0))
    path_count = problem3.count_all_paths_dfs()
    print(f"Total number of different paths to reach goal state: {path_count}")
    
    # Problem 2: Different goal state (2L in 3L jug)
    print("\n\n" + "="*60)
    print("Problem 2: 3L and 4L jugs, Goal: (2, 0)")
    print("-" * 60)
    
    problem4 = WaterJugProblem(jug1_capacity=3, jug2_capacity=4, 
                                goal_state=(2, 0), initial_state=(0, 0))
    solution_bfs2 = problem4.bfs()
    if solution_bfs2:
        problem4.print_solution(solution_bfs2, "BFS")
    else:
        print("No solution found")
    
    # Problem 3: Different capacities
    print("\n\n" + "="*60)
    print("Problem 3: 5L and 3L jugs, Goal: (4, 0)")
    print("-" * 60)
    
    problem5 = WaterJugProblem(jug1_capacity=5, jug2_capacity=3, 
                                goal_state=(4, 0), initial_state=(0, 0))
    solution_bfs3 = problem5.bfs()
    if solution_bfs3:
        problem5.print_solution(solution_bfs3, "BFS")
    else:
        print("No solution found")


if __name__ == "__main__":
    main()

File: Assignment-2/c/water_jug_problem.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include "water_jug_problem.h"

#define MAX_CAP 100

JugNode* create_node(int j1, int j2, const char* action, JugNode* parent) {
    JugNode* node = (JugNode*)malloc(sizeof(JugNode));
    node->state.jug1 = j1;
    node->state.jug2 = j2;
    strncpy(node->action, action, 49);
    node->action[49] = '\0';
    node->parent = parent;
    return node;
}

bool is_goal(JugState state, int t1, int t2) {
   
    return state.jug1 == t1 && state.jug2 == t2;
}

JugNode* solve_water_jug_bfs(int cap1, int cap2, int target1, int target2) {
    bool visited[MAX_CAP][MAX_CAP];
    memset(visited, 0, sizeof(visited));

    JugNode* queue[10000];
    int front = 0, rear = 0;

    JugNode* start = create_node(0, 0, "Initial State", NULL);
    queue[rear++] = start;
    visited[0][0] = true;

    while (front < rear) {
        JugNode* current = queue[front++];
        int j1 = current->state.jug1;
        int j2 = current->state.jug2;

        if (is_goal(current->state, target1, target2)) {
            return current;
        }

        if (j1 < cap1) {
            if (!visited[cap1][j2]) {
                visited[cap1][j2] = true;
                queue[rear++] = create_node(cap1, j2, "Fill Jug1", current);
            }
        }
        if (j2 < cap2) {
            if (!visited[j1][cap2]) {
                visited[j1][cap2] = true;
                queue[rear++] = create_node(j1, cap2, "Fill Jug2", current);
            }
        }
        if (j1 > 0) {
            if (!visited[0][j2]) {
                visited[0][j2] = true;
                queue[rear++] = create_node(0, j2, "Empty Jug1", current);
            }
        }
        if (j2 > 0) {
            if (!visited[j1][0]) {
                visited[j1][0] = true;
                queue[rear++] = create_node(j1, 0, "Empty Jug2", current);
            }
        }
        if (j1 > 0 && j2 < cap2) {
            int amount = (j1 < (cap2 - j2)) ? j1 : (cap2 - j2);
            int nj1 = j1 - amount;
            int nj2 = j2 + amount;
            if (!visited[nj1][nj2]) {
                visited[nj1][nj2] = true;
                queue[rear++] = create_node(nj1, nj2, "Pour Jug1->Jug2", current);
            }
        }
        if (j2 > 0 && j1 < cap1) {
            int amount = (j2 < (cap1 - j1)) ? j2 : (cap1 - j1);
            int nj1 = j1 + amount;
            int nj2 = j2 - amount;
            if (!visited[nj1][nj2]) {
                visited[nj1][nj2] = true;
                queue[rear++] = create_node(nj1, nj2, "Pour Jug2->Jug1", current);
            }
        }
    }
    return NULL;
}

JugNode* solve_water_jug_dfs(int cap1, int cap2, int target1, int target2) {
    JugNode* stack[10000];
    int top = 0;

    bool visited[MAX_CAP][MAX_CAP];
    memset(visited, 0, sizeof(visited));

    JugNode* start = create_node(0, 0, "Initial State", NULL);
    stack[top++] = start;
    
  
    
    while (top > 0) {
        JugNode* current = stack[--top];
        int j1 = current->state.jug1;
        int j2 = current->state.jug2;

        if (visited[j1][j2]) {
            continue; 
        }
        visited[j1][j2] = true;

        if (is_goal(current->state, target1, target2)) {
            return current;
        }

        
        int next_states[6][2];
        char actions[6][50];
        int count = 0;


        next_states[count][0] = cap1; next_states[count][1] = j2; strcpy(actions[count++], "Fill Jug1");
        next_states[count][0] = j1; next_states[count][1] = cap2; strcpy(actions[count++], "Fill Jug2");
        next_states[count][0] = 0; next_states[count][1] = j2; strcpy(actions[count++], "Empty Jug1");
        next_states[count][0] = j1; next_states[count][1] = 0; strcpy(actions[count++], "Empty Jug2");
        {
            int amt = (j1 < (cap2 - j2)) ? j1 : (cap2 - j2);
            next_states[count][0] = j1 - amt; next_states[count][1] = j2 + amt; strcpy(actions[count++], "Pour Jug1->Jug2");
        }
        {
            int amt = (j2 < (cap1 - j1)) ? j2 : (cap1 - j1);
            next_states[count][0] = j1 + amt; next_states[count][1] = j2 - amt; strcpy(actions[count++], "Pour Jug2->Jug1");
        }

        for (int i = 0; i < count; i++) {
            int nj1 = next_states[i][0];
            int nj2 = next_states[i][1];
            if (nj1 >= 0 && nj1 <= cap1 && nj2 >= 0 && nj2 <= cap2) {
                if (!visited[nj1][nj2]) {
                    stack[top++] = create_node(nj1, nj2, actions[i], current);
                }
            }
        }
    }
    return NULL;
}

void dfs_count(int j1, int j2, int cap1, int cap2, int t1, int t2, 
               bool path_visited[MAX_CAP][MAX_CAP], int depth, int max_depth, int* count) {
    
    if (j1 == t1 && j2 == t2) {
        (*count)++;
        return;
    }

    if (depth >= max_depth) return;

    path_visited[j1][j2] = true;

    int next_states[6][2];
    int c = 0;
    next_states[c][0] = cap1; next_states[c++][1] = j2;
    next_states[c][0] = j1; next_states[c++][1] = cap2;
  
    next_states[c][0] = 0; next_states[c++][1] = j2;
  
    next_states[c][0] = j1; next_states[c++][1] = 0;
 
    {
        int amt = (j1 < (cap2 - j2)) ? j1 : (cap2 - j2);
        next_states[c][0] = j1 - amt; next_states[c++][1] = j2 + amt;
    }

    {
        int amt = (j2 < (cap1 - j1)) ? j2 : (cap1 - j1);
        next_states[c][0] = j1 + amt; next_states[c++][1] = j2 - amt;
    }

    for (int i = 0; i < c; i++) {
        int nj1 = next_states[i][0];
        int nj2 = next_states[i][1];
        if (nj1 >= 0 && nj1 <= cap1 && nj2 >= 0 && nj2 <= cap2) {
            if (!path_visited[nj1][nj2]) {
                dfs_count(nj1, nj2, cap1, cap2, t1, t2, path_visited, depth + 1, max_depth, count);
            }
        }
    }

    path_visited[j1][j2] = false; 
}

int count_all_paths_dfs(int cap1, int cap2, int target1, int target2) {
    bool path_visited[MAX_CAP][MAX_CAP];
    memset(path_visited, 0, sizeof(path_visited));
    int count = 0;
    dfs_count(0, 0, cap1, cap2, target1, target2, path_visited, 0, 20, &count);
    return count;
}

void print_water_jug_solution(JugNode* goal, const char* algo_name, int cap1, int cap2, JugState initial) {
    if (!goal) return;

    // Reconstruct path
    JugNode* path[100]; // Max path
    int len = 0;
    JugNode* curr = goal;
    while (curr != NULL) {
        path[len++] = curr;
        curr = curr->parent;
    }

    // Print
    printf("\n============================================================\n");
    printf("Solution using %s\n", algo_name);
    printf("============================================================\n");
    printf("Initial State: (%d, %d)\n", initial.jug1, initial.jug2);
    printf("Goal State: (%d, %d)\n", goal->state.jug1, goal->state.jug2);
    printf("Jug Capacities: Jug1=%dL, Jug2=%dL\n", cap1, cap2);
    printf("\nSteps to reach goal:\n");
    printf("%-6s %-20s %-8s %-8s\n", "Step", "Action", "Jug1", "Jug2");
    printf("------------------------------------------------------------\n");

    // Python style:
    // Step 0: Initial State, Initial State
    // Step i: Action i, State i-1
    // Final Step: "Goal Reached", Goal State
    
    printf("%-6d %-20s %-8d %-8d\n", 0, "Initial State", initial.jug1, initial.jug2);
    
    for (int i = len - 1; i > 0; i--) {
        int step = len - i;
        // Print current node's action and its parent's state
        printf("%-6d %-20s %-8d %-8d\n", step, path[i-1]->action, path[i]->state.jug1, path[i]->state.jug2);
    }
    
    // Final goal reached step
    printf("%-6d %-20s %-8d %-8d\n", len, "Goal Reached", goal->state.jug1, goal->state.jug2);

    printf("\nTotal steps: %d\n", len);
}
